### ISSUE TITLE
CmsSigner.Sign() Prefix not injecting __instance correctly on Linux.

### ISSUE THREAD

**Issue by Fasjeit** [pardeike/Harmony#487]
## **Description**
`CmsSigner.Sign()` Prefix not injecting `__instance` correctly on Linux. (working on Windows).

**Prefix**
```csharp
//internal SignerInfoAsn Sign(ReadOnlyMemory<byte> data, string? contentTypeOid, bool silent, out X509Certificate2Collection chainCerts)
[HarmonyPatch("System.Security.Cryptography.Pkcs.CmsSigner", "Sign")]
internal class CmsSignerDetour
{
    public static bool Prefix(
       ReadOnlyMemory<byte> data,
        string? contentTypeOid,
        bool silent,
        out X509Certificate2Collection chainCerts,
        CmsSigner __instance,
        ref object __result) // SignerInfoAsn
    {
        FileLog.Log("Detoured!");
         // throws null-ref on Linux here, 
         // debugger shows "unable to read memory" for `__instance`
        var digestAlgorithm = __instance.DigestAlgorithm.Value;
        FileLog.Log($"digestAlgorithm is [{digestAlgorithm}]");

        chainCerts = null;
        return true;
    }
}
```
**Expected behavior**
```
Detoured!
digestAlgorithm is [2.16.840.1.101.3.4.2.1]
```

**Actual behavior (Linux)**
```
Detoured!
System.NullReferenceException: Object reference not set to an instance of an object.
   at Test.Security.Cryptography.Pkcs.Detours.CmsSignerDetour.Prefix(ReadOnlyMemory`1 data, String contentTypeOid, Boolean silent, X509Certificate2Collection& chainCerts, CmsSigner __instance, Object& __result) in C:\Users\makarov\Documents\Visual Studio 2013\Projects\TestOnly\HarmonyBugTest\CmsSignerDetour.cs:line 21
   at System.Security.Cryptography.Pkcs.CmsSigner.Sign_Patch1(CmsSigner this, ReadOnlyMemory`1 data, String contentTypeOid, Boolean silent, X509Certificate2Collection& chainCerts)
   at System.Security.Cryptography.Pkcs.SignedCms.ComputeSignature(CmsSigner signer, Boolean silent)
   at System.Security.Cryptography.Pkcs.SignedCms.ComputeSignature(CmsSigner signer)
   at HarmonyBugTest.Program.Main(String[] args) in C:\Users\X\Documents\Visual Studio 2013\Projects\TestOnly\HarmonyBugTest\Program.cs:line 188
```

## **Environment**
```
Linux Mint 19.3 Tricia 
dotnet sdk 6.0.101
Microsoft.NETCore.App 6.0.1
Harmony 2.2.1
```

## **To Reproduce**
1. Download attached zip archive (containing sample project) to linux machine.
[HarmonyBugTest.zip](https://github.com/pardeike/Harmony/files/8965832/HarmonyBugTest.zip)
2. `dotnet run`
3. see `harmony.log.txt`

## **Screenshots**
![Снимок](https://user-images.githubusercontent.com/15075638/175263408-f789d99c-fc63-48f6-a851-80d51c5256c9.PNG)

## **Harmony log**
[harmony.log.txt](https://github.com/pardeike/Harmony/files/8965849/harmony.log.txt)

**Comment by pardeike** [comment]
The problem lies within these heuristics:
https://github.com/pardeike/Harmony/blob/master/Harmony/Internal/StructReturnBufferCheck.cs

The return type of the method you patch is a struct `SignerInfoAsn` and on certain platforms and under certain conditions that are not documented, the JIT compiler does not use the stack for the returned result. Instead the calling convention is changed to that the caller allocates the result and passes a pointer to it as the second argument (first one is still “this”).

Harmony builds dynamic methods and has to fake the argument order to adapt to these calling conventions because dynamic methods won’t automatically get them. 

I assume that for some unknown reason the heuristics fail and Harmony uses the wrong use case.

**Comment by Fasjeit** [comment]
We are using harmony a lot in 6.0 both win and linux, having no issues beside this so far.

**Comment by Fasjeit** [comment]
Yes, looks like the problem is exactly in `GetStructReplacement_Net` for linux
https://github.com/pardeike/Harmony/blob/d420fd5b318b2f3d234b13edb59fc169b351df8c/Harmony/Internal/StructReturnBufferCheck.cs#L53-L65

The method behaves like there is no need for extra ptr in params, putting magic value in `a` and `ptr` args, instead of `a` and `b`.

![Снимок](https://user-images.githubusercontent.com/15075638/175531852-93af7651-1327-4e5e-b79b-844beb773602.PNG)

Changing the return value of `GetStructReplacement_Net` to `true` manually during debug fixes the problem.

ps. Also checked `net5.0` - the problem persists.

**Comment by pardeike** [comment]
The question is more if the heuristics that test the behaviour are correct designed for Linux. I don’t have a Linux distribution handy to test with but I suspect it has something to do with the number of bytes a struct takes up or the way the size is determined. 

I assume that if we can create the correct runtime test that flags the case correctly under Linux we can fix this. But the code need to work for all environments or have conditional code that checks specifically for Linux.

**Comment by Fasjeit** [comment]
Yes, I understand we need some general way to solve the problem. I am just logging my findings so far.

I've done a little bit of experimenting. Here is the setup and results.

#### 1 - replaced `SomeStruct_Net` with struct with array of selected size `N`
```csharp
internal unsafe struct SomeStruct_Net
{
#pragma warning disable CS0169
    public fixed byte headerBytes[N];
#pragma warning restore CS0169
}
```
#### 2 - changed `N` and observe behavior of `GetStructReplacement_Net`.

#### Win version works as expected

For all values of `N` except `1, 2, 4, 8` `GetStructReplacement_Net` returns `true`, i.e. we have return buffer pointer.

#### Linux version seems a little bit different.

At small values `GetStructReplacement_Net` returns `false` (i.e. compiler do not use struct buffer pointer in args), but after some threshold (>= 17) it starts to return `true`. 

My guess is the solution could be to add an extra check for linux platform at `NeedsFix` to check whether the return struct has the size more than 16 bytes, as it looks like compiler is not using return pointers for small structs there.

I'll double check and investigate a little bit more of this.

Also I have a question - is the creation of `StructReturnBuffer` serves any purpose here? It seems not used anywhere....
```csharp
lock (hasTestResult_Net_lock)
{
  if (hasTestResult_Net is false)
  {
    Sandbox.hasStructReturnBuffer_Net = false;
    var self = new StructReturnBuffer(); // ????????
    var original = AccessTools.DeclaredMethod(typeof(Sandbox), nameof(Sandbox.GetStruct_Net));
    var replacement = AccessTools.DeclaredMethod(typeof(Sandbox), nameof(Sandbox.GetStructReplacement_Net));
    _ = Memory.DetourMethod(original, replacement);
    _ = new Sandbox().GetStruct_Net(Sandbox.magicValue, Sandbox.magicValue);
    hasTestResult_Net = true;
  }
}
```

**Comment by pardeike** [comment]
First, awesome and well needed work!
Second, I’m unsure about that unused assignment but I’ll have a look at the blame history when I have time. 

/Andreas

**Comment by pardeike** [comment]
One more thing: I know of experience that this could be different if one would use .Net/Mono and 32/64Bit environments. I guess processor type (arm/x86/x64) probably too. I have to think about how to create tests for all cases.

**Comment by Fasjeit** [comment]
Can you please tell how platform/architecture/compiler checks is being done in Harmony, or where can I find some examples of these type of conditional checks in the code?

**Comment by pardeike** [comment]
Memory.cs uses `isWindows` which if false indicates Linux and macOS machines:
```cs
static readonly bool isWindows = Environment.OSVersion.Platform.Equals(PlatformID.Win32NT);
```

**Comment by pardeike** [comment]
For now, I did what you suggested and added a check `isWindows == false && size > 16` that returns `true` in `NeedsFix`

**Comment by Fasjeit** [comment]
Thank you, it looks like my use case now working correctly.

Still I don't quite understand why we just omit `GetStructReplacement_Net` check for linux and keep it for windows. Maybe I just don't understand why we need it in the fist place... 

My understanding is in some cases even for sizes except 1, 2, 4 and 8 we can still don't have struct return pointer, so we check it empirically during runtime using struct size of 3 bytes for windows. Should we still have this check for linux, but with 17 byte struct, instead of 3?

Maybe it is better to do something like this? (Create separate struct and method for linux):
see commit here: https://github.com/Fasjeit/Harmony/commit/b0ff5ba921652a3a8f4b02752ef6d370a60b0576, it works as expected.

Or maybe I am just wrong...

**Comment by pardeike** [comment]
fixed in d1aa13b919adf1089ec8b3198c5c081873c51440 and part of the 2.2.2 release
